### ISSUE TITLE
Generation never stops

### ISSUE THREAD

**Issue by ortegaalfredo** [turboderp/exllamav2#9]
I'm testing it with TheBloke's airoboros-70b-2.1 in a 2x3090. Generating with the simple generator works great, output is the same quality but much faster than exllama-v1, but I cannot make it to stop generating. It just continue generating until it hit max_new_len tokens. The line:

                settings.disallow_tokens(tokenizer, [tokenizer.eos_token_id])

Makes no difference.

UPDATE: this fixes it:

```
diff --git a/exllamav2/generator/base.py b/exllamav2/generator/base.py
index 63db5f5..c6ed9d6 100644
--- a/exllamav2/generator/base.py
+++ b/exllamav2/generator/base.py
@@ -58,6 +58,8 @@ class ExLlamaV2BaseGenerator:
 
             logits = self.model.forward(self.sequence_ids[:, -1:], self.cache, input_mask = mask).float().cpu()
             token, _ = ExLlamaV2Sampler.sample(logits, gen_settings, self.sequence_ids, random.random())
+            if token==self.tokenizer.eos_token_id:
+                break
             self.sequence_ids = torch.cat([self.sequence_ids, token], dim = 1)
 
         text = self.tokenizer.decode(self.sequence_ids)

```

**Comment by ScottSump** [comment]
mines doing the same thing, same settings.

**Comment by turboderp** [comment]
I'm not sure what the intended prompt format is for Airoboros. But you typically need to get that exactly right to make the model emit a properly formatted response as well, EOS token and all.

**Comment by ortegaalfredo** [comment]
But I think there is a bug in the example code. It never detect the eos_token at all, even if the LLM emits it. I updated the issue with a patch that works. Still the model sometimes gets out of control with some prompts. It seem to be very sensitive to temperature, unlike exllama v1.

**Comment by Qubitium** [comment]
@ortegaalfredo  We fixed the never-ending generation in v2 by adding/prepending BOS ```<s>``` to the prompt via code: New v2 tokenizer doesn't have the toggle to encode special tokens so it throws away BOS from our prompt at this point, causing the never ending generation for our specific model. 

```
self.tokenizer.encode(prompt, add_bos=True)
```

Fyi, our model is trained where prompts always start with ```<s>``` so this may not apply to your situation. In any case, print out the tokens after encode to maker sure everything is matching v1 code.

**Comment by turboderp** [comment]
Yeah, I still need an elegant solution to this. I'll add the option to encode control symbols at some point, but I kind of understand what the creators of SentencePiece were thinking when they decided not to allow it (without a lot of jumping through hoops.) 

It technically requires sanitizing all user input with regards to rules that could change from one model to the next, and since encoding is ambiguous you have to impose special rules to make sure no other tokens include the control tokens. For instance a training dataset might have `\n</s>` appear many times, but that string should never be a token anyway.

**Comment by Rajmehta123** [comment]
Any updates on this fix?

**Comment by turboderp** [comment]
Yes.

You can still use `add_bos` when tokenizing to prepend BOS to the the token IDs, or `add_eos` to append EOS.

Alternatively, if you encode with `encode_special_tokens = True`, any control symbols in the input string will be encoded to their respective token IDs. Likewise, you can decode with `decode_special_tokens = True` to turn control tokens back into strings.

In other words:

```python
tokenizer.encode("Hello", add_bos = True, add_eos = True)
-> [tokenizer.bos_token_id] + tokenizer.encode("Hello") + [tokenizer.eos_token_id]

tokenizer.encode("<s>Hello</s>", encode_special_tokens = True)
-> [tokenizer.bos_token_id] + tokenizer.encode("Hello") + [tokenizer.eos_token_id]
```

The latter will also respect extra tokens defined in `added_tokens.json`, like the `<|im_start|>` etc. used by ChatML models.

**Comment by Rajmehta123** [comment]
What changes to make? add_bos and eos to generate base.py is actually not working. 

```
### EXLLAMA V2
model_directory =  "/home/ec2-user/.gccc/text-generation-webui/models/TheBloke_Airoboros-L2-70b-2.2.1-GPTQ"
import sys, os


import os,sys


os.chdir('/home/ec2-user/.gccc/exllamav2')
from exllamav2 import(
    ExLlamaV2,
    ExLlamaV2Config,
    ExLlamaV2Cache,
    ExLlamaV2Tokenizer,
)

from exllamav2.generator import (
    ExLlamaV2BaseGenerator,
    ExLlamaV2Sampler
)

import time

# Initialize model and cache



config = ExLlamaV2Config()
config.model_dir = model_directory
config.prepare()

model = ExLlamaV2(config)
print("Loading model: " + model_directory)

# allocate 18 GB to CUDA:0 and 24 GB to CUDA:1.
# (Call `model.load()` if using a single GPU.)
model.load([20,20,24,24])

tokenizer = ExLlamaV2Tokenizer(config)

cache = ExLlamaV2Cache(model)

# Initialize generator

generator = ExLlamaV2BaseGenerator(model, cache, tokenizer)
generator.warmup()
# Generate some text

settings = ExLlamaV2Sampler.Settings()
settings.temperature = 0.25
settings.top_k = 50
settings.top_p = 0.25
settings.token_repetition_penalty = 1.15
settings.disallow_tokens(tokenizer, [tokenizer.eos_token_id])


prompt_gen = f"""BEGININPUT
BEGINCONTEXT
#page: 16
ENDCONTEXT
02. MARKET OVERVIEW \nyA Ay> COMPETITIVE LANDSCAPE with the large software providers compete with the large software providers is a challenger platform with the ability to compete with the large software providers \nBenchmarking strongly against the functionality of the established and well-funded products.
ENDINPUT
BEGININSTRUCTION
Don't make up answers if you don't know.
Question: {question}
ENDINSTRUCTION
"""

settings.temperature = 0.25
settings.top_p = 0.25
settings.top_k = 60
max_new_tokens=512
generation_output = generator.generate_simple(prompt_gen, settings, max_new_tokens)

#    generator.generate_simple(prompt, settings, max_new_tokens, seed = 1234)
ans = generation_output.split("ENDINSTRUCTION")[-1].split("BEGINANSWER")[-1].split("ENDANSWER")[0].strip()
print(ans)

```

Response:

```
Answer: The competitive landscape is quite strong as it can compete with other big players in the industry [Redacted]. However, some of these features are only partially available or have minimal implementation.

Reference(s):
Page 16, [https://www.dropbox.com/s/q379vkxbjw4fzgd/vega_prospectus.pdf?raw=1](https://www.dropbox.com/s/q379vkxbjw4fzgd/vega_prospectus.pdf?raw=1)
*/

```

Reference(s):
Page 16, https://www.dropbox.com/s/q379vkxbjw4fzgd/vega_prospectus.pdf?raw=1
```

---

### Alpha Financial Software Holdings plc - Investor Presentation - September 2018

BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
BEGININPUT
```

**Comment by Rajmehta123** [comment]
Also, if I run in text generation web UI, the generations stop at newline if I keep the EOS token. But if I ban the eos token, the output is generated until max_new_tokens. Is it a problem with the model that it is not generating the EOS token for it to stop?

**Comment by Rajmehta123** [comment]
Okay. I might know what the problem is. The eos token ID is </s> for the vicuna-style prompt. But the airoboros also supports closed context prompts. The prompt format is changing and that prompt style is not generating </s>. Hence, the eos token id works for the vicuna style but not for the closed context style. Please correct me if I am wrong. @turboderp

**Comment by turboderp** [comment]
I'm not clear on what the prompt format is or why there are two of them? But you could consider using the streaming generator instead, as it supports multiple different stop conditions (including stop strings) and gives you a bit more control since it starts the generation from a set of token IDs instead.

**Comment by RisaKirisu** [comment]
@turboderp I observed the same issue when using Llama2-70b-chat. Exl2 format 2.3bpw. The eos token id for llama2 is 2. I'm used the correct prompt format and verified that bos token is added to my prompt by printing the encoded token_ids out. The issue is that when an eos token is generated by the model, ```ExLlamaV2Sampler.sample``` doesn't return ```eos``` as ```True```. I printed the tokens generated and eos returned by the sampler and found out this is the case:
```
for i in range(num_tokens):

    logits = self.model.forward(self.sequence_ids[:, -1:], self.cache, input_mask = mask, loras = loras).float().cpu()
    token, _, eos = ExLlamaV2Sampler.sample(logits, gen_settings, self.sequence_ids, random.random(), self.tokenizer, prefix_token = unhealed_token)
    self.sequence_ids = torch.cat([self.sequence_ids, token], dim = 1)
    gen_settings.feed_filters(token)

    print("token: {}, eos:{}".format(token[0].item(), eos))
    unhealed_token = None
    if eos: break
```
![image](https://github.com/turboderp/exllamav2/assets/28692567/803ba1bd-7a91-4881-af06-11c7e2d9db12)

The returned ```eos``` is always false, and thus the generation look never stop until max output token length is reached.

In ```ExLlamaV2Sampler.sample```, this return value is only related to the filter, and I don't know what a filter is:
![image](https://github.com/turboderp/exllamav2/assets/28692567/d89cd7c3-26f0-4387-adba-d64427345a6d)

If the correct behavior for generator is to stop generating when an eos token is generated, then perhaps the stop condition ```if eos: break``` is wrong since the value of this ```eos``` has nothing to do with eos token.

**Comment by RisaKirisu** [comment]
but if breaking the loop when eos is detected like OP did only work for single prompt input. For batch inference is it always necessary to generate all ```num_tokens``` tokens for every prompt?

**Comment by turboderp** [comment]
Well, this was an oversight on my part. I was too focused on the streaming generator when updating the sampling logic and messing around with grammar and stuff, so `generate_simple` was overlooked in all that.

Indeed the `eos` signal from the sampler has a different meaning and relates to sampling filters (full implementation still in the works), with the actual EOS handling being delegated to the generator loop.

To make matters more confusing, SentencePiece was masking the issue by only decoding up the EOS token anyway.

The latest commit should fix it, though. You do need to handle batches, yes, so if batching it will keep generating until every sequence in the batch has reached an EOS token, replacing any tokens after EOS with padding tokens that are then decoded to nothing.

**Comment by chaseaucoin** [comment]
I'm having this same issue when using this model: https://huggingface.co/JCTN/Mistral-7B-instruct-exl2 and the streaming example that leverages ExLlamaV2StreamingGenerator from the examples folder with the sample prompt from the hugging face page.

prompt = """<s>[INST] What is your favourite condiment? [/INST]
Well, I'm quite partial to a good squeeze of fresh lemon juice. It adds just the right amount of zesty flavour to whatever I'm cooking up in the kitchen!</s>
[INST] Do you have mayonnaise recipes? [/INST]"""

I did add the encode_special_tokens as mentioned in this thread:

input_ids = tokenizer.encode(prompt, encode_special_tokens = True)

I have tried with and without

settings.disallow_tokens(tokenizer, [tokenizer.eos_token_id])

I am uncertain what I should change in the main loop to get it to end properly.

while True:
    chunk, eos, _ = generator.stream()
    generated_tokens += 1
    print (chunk, end = "")
    sys.stdout.flush()
    if eos or generated_tokens == max_new_tokens: break

**Comment by turboderp** [comment]
The prompt format isn't quite right there. Here's what I use in ExUI, for reference, to produce one prompt-response round:

```python
    def format(self, prompt, response, system_prompt, settings):
        text = "<s>[INST] "
        if system_prompt and system_prompt.trim() != "":
            text += "<<SYS>>\n"
            text += system_prompt
            text += "\n<</SYS>>\n\n "
        text += prompt
        text += " [/INST]"
        if response:
            text += response
            text += "</s>"
        return text
```

Note that `<s>` appears in every round, and `</s>` is part of the model's response only. Maybe less important but still worth noting is that the first token of the response will typically (always?) start with a space. So if you're re-encoding a response where that leading space has been trimmed off, you should add it back in. 

Also note that `encode_special_tokens` doesn't use the lstrip/rstrip options from Transformers, so `<s>` and `</s>` here are just delimiters and any whitespace immediately before or after those symbols will be tokenized. I.e. `<s> Hello` and `<s>Hello` will encode differently.

All in all, as far as I understand the format, your example should be:

`<s>[INST] What is your favourite condiment? [/INST] Well, I'm quite partial to a good squeeze of fresh lemon juice. It adds just the right amount of zesty flavour to whatever I'm cooking up in the kitchen!</s><s>[INST] Do you have mayonnaise recipes? [/INST]`

**Comment by chaseaucoin** [comment]
@turboderp I tried your suggestion. Changing the format made zero difference. The example was pulled from https://huggingface.co/JCTN/Mistral-7B-instruct-exl2

**Comment by turboderp** [comment]
That's odd. And you're sure the stop condition is set?

```python
generator.set_stop_conditions([tokenizer.eos_token_id])
```

**Comment by chaseaucoin** [comment]
Yay! That combination did it! Thank you @turboderp you rock! 

For anyone else who runs into this I removed this from the example:

`settings.disallow_tokens(tokenizer, [tokenizer.eos_token_id])`

and added 

`generator.set_stop_conditions([tokenizer.eos_token_id])`

I'd be happy to update the examples or add a new one are you accepting PRs?

**Comment by turboderp** [comment]
Idk, there's a lot to prompting models. I guess I'll have to put some thought into a tutorial of some kind because there really isn't one canonical approach to prompt formats, stop conditions and prompting strategy. And as long as new models keep coming out every other day with their own particular quirks, it's really hard to stay ahead of all that.

**Comment by nxitik** [comment]
I am trying with yarn mistral 7B-64k GPTQ and i am having issues that @chaseaucoin mentioned about the model filling in garbage after the answer to fill up the max_new_tokens. Any help?

**Comment by turboderp** [comment]
Well, Yarn isn't supported directly. You can use RoPE scaling by setting the parameter in the model config, and here I guess the scaling factor would be 8? Whether that exactly replicates the behavior that the Yarn people decided to add to the "standard", I'm not sure. It's impossible to keep up with what so many other people are adding into the mix. 

If the model is working anyway, it may not be working optimally. But in any case, you need to make sure the prompt format is correct and you have appropriate stop conditions set for that format.
